### ISSUE TITLE
`sqlmesh plan` attempts to create physical tables even when they exist, adding unnecessary api calls wait time

### ISSUE THREAD

**Issue by esciara** [TobikoData/sqlmesh#3336]
For example, using BigQuery with having 59 models makes that `sqlmesh plan` will always take an additional minute, even though one table is changed and only one physical table needs creation :

<img width="734" alt="Screenshot 2024-11-06 at 13 25 35" src="https://github.com/user-attachments/assets/0d6c36d0-82f8-4cf2-8fb1-4590a9d82cb0">

Also there are a lot of calls for just a `SELECT 1`, presumably to check connection works.

<img width="1025" alt="image" src="https://github.com/user-attachments/assets/d7a557ca-041e-4841-91b8-f4344584825b">

The unnecessary calls and API response waits make iterative work difficult, especially when it is not possible to do local development with a different database engine (DuckDB for instance).

The bottom line for this issue is: how to optimize api calls to call less for unnecessary SQL jobs.

**Comment by erindru** [comment]
I believe this is done mostly for "self healing". There are a bunch of places where SQLMesh can fail / the process exits and it needs to be able to pick up reliably where it left off on the next invocation.

However, there is definite room for improvement here. Perhaps we could track which tables exist within a single invocation of sqlmesh (and bust the cache if theyre dropped) to prevent extra queries getting issued. I can see this leading to a bunch of cache-related edge cases though.

With regards to the `SELECT 1` queries, [theyre used](https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/engine_adapter/bigquery.py#L156) to establish a [BigQuery session](https://cloud.google.com/bigquery/docs/sessions-intro) which allows us to do things like create temporary tables and have them isolated from other sessions and automatically dropped at the end of the session to prevent garbage accumulating

**Comment by esciara** [comment]
Thanks for the explanations. I would love to help on this. If you could share some pointers to where to look, that would help me getting to where I could change things more quickly.

**Comment by esciara** [comment]
was on `0.129.0`. and indeed, it is now waaaaaaay faster (upgraded to lastest, meaning `0.130.2`).
